File: pkg/material/checker_sphere.go
package material

import (
	"math"
	"math/rand"

	"sif/gogs/eric/relray/pkg/geometry"
	"sif/gogs/eric/relray/pkg/spectrum"
	"sif/gogs/eric/relray/pkg/vec"
)

// CheckerSphere is a Lambertian diffuse material with a checkerboard pattern
// mapped onto a sphere using latitude/longitude coordinates derived from the
// surface normal. NumSquares controls how many checker divisions there are
// around the equator (and half that many pole-to-pole).
type CheckerSphere struct {
	Even, Odd  spectrum.SPD
	NumSquares int // divisions around equator (default 8)
}

func (c *CheckerSphere) Scatter(inDir vec.Vec3, hit geometry.Hit, rng *rand.Rand) ScatterResult {
	scattered := hit.Normal.Add(RandomUnitVec(rng))
	if scattered.LengthSq() < 1e-12 {
		scattered = hit.Normal
	}
	return ScatterResult{
		Scattered:   true,
		OutDir:      scattered.Normalize(),
		Reflectance: c.reflectanceAt(hit.Normal),
	}
}

func (c *CheckerSphere) Emitted(hit geometry.Hit) spectrum.SPD {
	return spectrum.SPD{}
}

func (c *CheckerSphere) reflectanceAt(normal vec.Vec3) spectrum.SPD {
	n := c.NumSquares
	if n <= 0 {
		n = 8
	}

	// Latitude: -π/2 to π/2, longitude: -π to π
	lat := math.Asin(clamp(normal.Y, -1, 1))
	lon := math.Atan2(normal.Z, normal.X)

	// Map to grid squares
	latDiv := int(math.Floor(lat * float64(n) / math.Pi))
	lonDiv := int(math.Floor(lon * float64(n) / math.Pi))

	if (latDiv+lonDiv)%2 == 0 {
		return c.Even
	}
	return c.Odd
}

func clamp(v, lo, hi float64) float64 {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

File: rust/src/material/checker_sphere.rs
use std::f64::consts::PI;

use crate::geometry::Hit;
use crate::spectrum::Spd;
use crate::vec::Vec3;
use rand::rngs::SmallRng;

use super::{random_unit_vec, Material, ScatterResult};

/// Lambertian diffuse material with a checkerboard pattern mapped onto a
/// sphere using latitude/longitude coordinates derived from the surface
/// normal. `num_squares` controls how many checker divisions there are
/// around the equator (and half that many pole-to-pole).
pub struct CheckerSphere {
    pub even: Spd,
    pub odd: Spd,
    /// Divisions around equator (default 8).
    pub num_squares: i32,
}

impl CheckerSphere {
    fn reflectance_at(&self, normal: Vec3) -> Spd {
        let n = if self.num_squares <= 0 { 8 } else { self.num_squares };

        // Latitude: -PI/2 to PI/2, longitude: -PI to PI
        let lat = normal.y.clamp(-1.0, 1.0).asin();
        let lon = normal.z.atan2(normal.x);

        // Map to grid squares
        let lat_div = (lat * f64::from(n) / PI).floor() as i64;
        let lon_div = (lon * f64::from(n) / PI).floor() as i64;

        if (lat_div + lon_div) % 2 == 0 {
            self.even
        } else {
            self.odd
        }
    }
}

impl Material for CheckerSphere {
    fn scatter(&self, _in_dir: Vec3, hit: &Hit, rng: &mut SmallRng) -> ScatterResult {
        let mut scattered = hit.normal + random_unit_vec(rng);
        if scattered.length_sq() < 1e-12 {
            scattered = hit.normal;
        }
        ScatterResult {
            scattered: true,
            out_dir: scattered.normalize(),
            reflectance: self.reflectance_at(hit.normal),
        }
    }

    fn emitted(&self, _hit: &Hit) -> Spd {
        Spd::default()
    }
}
